File: app/src/main/java/com/airadar/app/data/BackendClient.kt
package com.airadar.app.data

import com.airadar.app.BuildConfig
import org.json.JSONArray
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.MediaType.Companion.toMediaType
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.sync.Mutex
import android.os.Build
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.io.IOException
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.format.DateTimeParseException
import java.util.concurrent.TimeUnit

/**
 * The Airadar backend (backend/ in this repo). It is the app's only flight-status
 * source: the AirLabs key stays on the server, and the wire shape mirrors [Flight]
 * field for field, so nothing here has to know which provider answered.
 */
object BackendClient {

    private val http = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    class BackendException(message: String, val code: Int = 0) : IOException(message)

    val isConfigured: Boolean get() = BuildConfig.BACKEND_URL.isNotBlank()

    /** Status for one flight number on one date; the server picks live vs timetable. */
    suspend fun flight(flightNumber: String, date: LocalDate): Flight = withContext(Dispatchers.IO) {
        val number = flightNumber.uppercase()
        val path = "flights/$number/$date"
        // Signed in, the server also knows the plan: Premium reaches past flights.
        val row = if (AuthStore.isSignedIn) runCatching { authed("GET", path) }.getOrElse { e ->
            if (e is BackendException && e.code == 401) call("GET", path) else throw e
        } else call("GET", path)
        parse(row, number, date)
    }

    /** Reference data for an airport the bundled table does not know. */
    suspend fun airport(iata: String): Airport = withContext(Dispatchers.IO) {
        val row = get("airports/${iata.uppercase()}")
        Airport(
            iata = row.getString("iata"),
            icao = row.optString("icao"),
            name = row.optString("name").ifBlank { iata.uppercase() },
            city = row.optString("city").ifBlank { iata.uppercase() },
            country = row.optString("country"),
            countryCode = row.optString("countryCode"),
            latitude = row.getDouble("latitude"),
            longitude = row.getDouble("longitude"),
            zoneId = row.optString("zoneId").ifBlank { "UTC" }
        )
    }

    // ---- account ------------------------------------------------------------

    /** Trades a Google ID token for this server's own session; stores it. */
    suspend fun signInWithGoogle(idToken: String): AuthUser = withContext(Dispatchers.IO) {
        val body = JSONObject().put("idToken", idToken).put("device", "${Build.MANUFACTURER} ${Build.MODEL}")
        storeSession(call("POST", "auth/google", body))
    }

    suspend fun signOut() = withContext(Dispatchers.IO) {
        AuthStore.refreshToken?.let { token ->
            runCatching { call("POST", "auth/logout", JSONObject().put("refreshToken", token)) }
        }
        AuthStore.clear()
    }

    private val refreshLock = Mutex()

    /**
     * Swaps the refresh token for a new pair. Serialised: two requests that both
     * hit a 401 must not both spend the same refresh token.
     */
    private suspend fun refreshSession(): Boolean {
        refreshLock.withLock {
            val token = AuthStore.refreshToken ?: return false
            return try {
                storeSession(call("POST", "auth/refresh", JSONObject().put("refreshToken", token)))
                true
            } catch (e: BackendException) {
                // Only a definite "no" ends the session; a dead network keeps it.
                if (e.code == 401) AuthStore.clear()
                false
            }
        }
    }

    private fun storeSession(json: JSONObject): AuthUser {
        val u = json.getJSONObject("user")
        val user = AuthUser(u.getString("id"), u.getString("email"), u.text("name"), u.text("avatarUrl"))
        AuthStore.save(json.getString("accessToken"), json.getString("refreshToken"), user)
        return user
    }

    // ---- trips ---------------------------------------------------------------

    suspend fun listTrips(deleted: Boolean = false): List<Flight> = withContext(Dispatchers.IO) {
        val arr = authed("GET", if (deleted) "trips/deleted" else "trips").getJSONArray("items")
        (0 until arr.length()).map { flightFromJson(arr.getJSONObject(it)) }
            .also { list -> list.forEach { ensureAirports(it.departure, it.arrival) } }
    }

    suspend fun putTrip(flight: Flight): Unit = withContext(Dispatchers.IO) {
        authed("PUT", "trips/${flight.id}", flight.toJson())
    }

    suspend fun deleteTrip(id: String): Unit = withContext(Dispatchers.IO) {
        authed("DELETE", "trips/$id")
    }

    suspend fun restoreTrip(id: String): Unit = withContext(Dispatchers.IO) {
        authed("POST", "trips/$id/restore")
    }

    // ---- people and shares -----------------------------------------------------

    suspend fun me(): AuthUser = withContext(Dispatchers.IO) { profile(authed("GET", "me")) }

    suspend fun updateProfile(findableByEmail: Boolean): AuthUser = withContext(Dispatchers.IO) {
        profile(authed("PATCH", "me", JSONObject().put("findableByEmail", findableByEmail)))
    }

    private fun profile(o: JSONObject): AuthUser {
        val current = AuthStore.user.value
        val membership = o.optJSONObject("membership")?.let(::membershipFromJson)
        AuthStore.updateProfile(o.text("givenName"), o.text("color"), o.optBoolean("findableByEmail", false), membership)
        return AuthUser(
            o.getString("id"), o.getString("email"), o.text("name") ?: current?.name, o.text("avatarUrl"),
            o.text("givenName"), o.text("color"), o.optBoolean("findableByEmail", false),
            membership ?: Membership.GUEST
        )
    }

    private fun membershipFromJson(o: JSONObject): Membership = Membership(
        tier = runCatching { Tier.valueOf(o.optString("tier").uppercase()) }.getOrDefault(Tier.GUEST),
        until = o.text("until")?.let { raw ->
            runCatching { java.time.OffsetDateTime.parse(raw).toInstant() }.getOrNull()
                ?: runCatching { LocalDateTime.parse(raw.take(19)).atOffset(java.time.ZoneOffset.UTC).toInstant() }.getOrNull()
        },
        trial = o.optBoolean("trial", false),
        token = o.text("token")
    )

    /** A token bought on the web page: binds it to this account (first come) and raises the plan. */
    suspend fun redeem(token: String): Membership = withContext(Dispatchers.IO) {
        val o = authed("POST", "billing/redeem", JSONObject().put("token", token.trim().uppercase()))
        membershipFromJson(o).also(AuthStore::saveMembership)
    }

    suspend fun lookup(email: String): Person = withContext(Dispatchers.IO) {
        personFromJson(authed("GET", "users/lookup?email=" + java.net.URLEncoder.encode(email.trim(), "UTF-8")))
    }

    suspend fun friends(): List<Friend> = withContext(Dispatchers.IO) {
        authed("GET", "friends").getJSONArray("items").let { arr -> (0 until arr.length()).map { friendFromJson(arr.getJSONObject(it)) } }
    }

    suspend fun requestFriend(userId: String): Friend = withContext(Dispatchers.IO) {
        friendFromJson(authed("POST", "friends/request", JSONObject().put("userId", userId)))
    }

    suspend fun acceptFriend(friendshipId: String): Friend = withContext(Dispatchers.IO) {
        friendFromJson(authed("POST", "friends/$friendshipId/accept"))
    }

    suspend fun removeFriend(friendshipId: String): Unit = withContext(Dispatchers.IO) {
        authed("DELETE", "friends/$friendshipId")
    }

    /** Shares with a friend; with no [toUserId] a link is minted and its URL returned. */
    suspend fun shareTrip(tripId: String, toUserId: String? = null): Pair<TripShare?, String?> =
        withContext(Dispatchers.IO) {
            val body = JSONObject()
            toUserId?.let { body.put("toUserId", it) }
            val o = authed("POST", "trips/$tripId/share", body)
            shareFromJson(o) to (o.text("url") ?: o.text("token")?.let { linkUrl(it) })
        }

    /** Friend shares of my trips: tripId → shares. */
    suspend fun outgoingShares(): Map<String, List<TripShare>> = withContext(Dispatchers.IO) {
        val arr = authed("GET", "shares/outgoing").getJSONArray("items")
        (0 until arr.length()).map { arr.getJSONObject(it) }
            .mapNotNull { o -> shareFromJson(o)?.let { o.getString("tripId") to it } }
            .groupBy({ it.first }, { it.second })
    }

    /** Friends' trips shared with me, each carrying who shared it and my answer so far. */
    suspend fun incomingShares(): List<Flight> = withContext(Dispatchers.IO) {
        val arr = authed("GET", "shares/incoming").getJSONArray("items")
        (0 until arr.length()).mapNotNull { i ->
            val o = arr.getJSONObject(i)
            val share = shareFromJson(o) ?: return@mapNotNull null
            val trip = flightFromJson(o.getJSONObject("trip"))
            ensureAirports(trip.departure, trip.arrival)
            trip.copy(id = "shared:${share.id}", sharedBy = share, deletedAt = null)
        }
    }

    suspend fun respondToShare(shareId: String, action: String): Unit = withContext(Dispatchers.IO) {
        authed("POST", "shares/$shareId/respond", JSONObject().put("action", action))
    }

    suspend fun unshare(shareId: String): Unit = withContext(Dispatchers.IO) {
        authed("DELETE", "shares/$shareId")
    }

    class LinkedTrip(val flight: Flight, val owner: Person, val ownerName: String)

    /** What a shared link points at; needs no sign-in. */
    suspend fun linkedTrip(token: String): LinkedTrip = withContext(Dispatchers.IO) {
        val o = call("GET", "shares/link/$token")
        val trip = flightFromJson(o.getJSONObject("trip"))
        ensureAirports(trip.departure, trip.arrival)
        val owner = personFromJson(o.getJSONObject("owner"))
        LinkedTrip(trip, owner, o.text("ownerName") ?: owner.givenName)
    }

    suspend fun copyLinkedTrip(token: String): Flight = withContext(Dispatchers.IO) {
        flightFromJson(authed("POST", "shares/link/$token/copy"))
    }

    fun linkUrl(token: String): String = BuildConfig.BACKEND_URL.trimEnd('/') + "/s/" + token

    private suspend fun ensureAirports(vararg codes: String) {
        codes.forEach { code -> FlightDatabase.ensureAirport(code) { airport(code) } }
    }

    // ---- transport --------------------------------------------------------

    private fun get(path: String): JSONObject = call("GET", path)

    /** A signed-in call: retried once with a fresh access token after a 401. */
    private suspend fun authed(method: String, path: String, body: JSONObject? = null): JSONObject {
        val first = AuthStore.accessToken ?: throw BackendException("Sign in required.", 401)
        return try {
            call(method, path, body, bearer = first)
        } catch (e: BackendException) {
            if (e.code != 401 || !refreshSession()) throw e
            call(method, path, body, bearer = AuthStore.accessToken)
        }
    }

    private val jsonMedia = "application/json; charset=utf-8".toMediaType()

    private fun call(method: String, path: String, body: JSONObject? = null, bearer: String? = null): JSONObject {
        val base = BuildConfig.BACKEND_URL.trimEnd('/')
        if (base.isBlank()) {
            throw BackendException("Backend address missing — set backend.url in local.properties and rebuild.")
        }
        // OkHttp insists on a body for POST/PUT, even an empty one, and none for GET.
        val requestBody = body?.toString()?.toRequestBody(jsonMedia)
            ?: if (method == "GET") null else ByteArray(0).toRequestBody(null)
        val request = Request.Builder()
            .url("$base/$path".toHttpUrl())
            .method(method, requestBody)
            .apply {
                BuildConfig.BACKEND_TOKEN.takeIf { it.isNotBlank() }?.let { header("X-Airadar-Token", it) }
                bearer?.let { header("Authorization", "Bearer $it") }
            }
            .build()

        val (code, text) = try {
            http.newCall(request).execute().use { it.code to it.body?.string().orEmpty() }
        } catch (e: IOException) {
            throw BackendException("Could not reach the Airadar server: ${e.message ?: e.javaClass.simpleName}")
        }

        if (code in 200..299) {
            // Lists arrive as a bare array; wrap so every caller sees an object.
            val trimmed = text.trim()
            return when {
                trimmed.startsWith("[") -> JSONObject().put("items", JSONArray(trimmed))
                trimmed.isEmpty() -> JSONObject()
                else -> JSONObject(trimmed)
            }
        }

        // FastAPI wraps errors as {"detail": ...}; ours carry an object with "error".
        val detail = runCatching { JSONObject(text).opt("detail") }.getOrNull()
        val reason = when (detail) {
            is JSONObject -> detail.optString("error").ifBlank { detail.toString() }
            null -> ""
            else -> detail.toString()
        }
        throw BackendException(
            when (code) {
                401 -> reason.ifBlank { "The server rejected this build's token (backend.token in local.properties)." }
                402 -> reason.ifBlank { "This needs a higher plan." }
                403 -> reason.ifBlank { "Not allowed." }
                404 -> if (path.startsWith("flights/")) "Nothing found for that flight on that date." else reason.ifBlank { "Not found." }
                429 -> "AirLabs monthly quota used up on the server."
                else -> reason.ifBlank { "Airadar server error (HTTP $code)." }
            },
            code
        )
    }

    // ---- parsing ----------------------------------------------------------

    private suspend fun parse(row: JSONObject, flightNumber: String, date: LocalDate): Flight {
        val dep = row.getString("departure").uppercase()
        val arr = row.getString("arrival").uppercase()
        // Any airport the server mentions can be placed: fetch and remember unknown ones.
        FlightDatabase.ensureAirport(dep) { airport(dep) }
        FlightDatabase.ensureAirport(arr) { airport(arr) }

        return Flight(
            id = row.optString("id").ifBlank { "$flightNumber-$date" },
            flightNumber = row.optString("flightNumber").ifBlank { flightNumber },
            airlineName = row.optString("airlineName").ifBlank { flightNumber.take(2) },
            departure = dep,
            arrival = arr,
            departureTerminal = row.text("departureTerminal"),
            arrivalTerminal = row.text("arrivalTerminal"),
            departureGate = row.text("departureGate"),
            arrivalGate = row.text("arrivalGate"),
            departureTime = row.time("departureTime")
                ?: throw BackendException("Record for $flightNumber has no scheduled departure."),
            arrivalTime = row.time("arrivalTime")
                ?: throw BackendException("Record for $flightNumber has no scheduled arrival."),
            status = runCatching { FlightStatus.valueOf(row.optString("status")) }
                .getOrDefault(FlightStatus.SCHEDULED),
            aircraft = row.text("aircraft"),
            baggageClaim = row.text("baggageClaim"),
            delayMinutes = row.optInt("delayMinutes", 0).coerceAtLeast(0),
            callsign = row.text("callsign")
                ?: FlightDatabase.airlineIcao(flightNumber.takeWhile { it.isLetter() })
                    ?.let { it + flightNumber.filter(Char::isDigit) }
        )
    }

    /** Server times are ISO local: `2026-09-21T02:30:00`. */
    private fun JSONObject.time(key: String): LocalDateTime? =
        text(key)?.let { raw ->
            try {
                LocalDateTime.parse(raw.take(19))
            } catch (_: DateTimeParseException) {
                null
            }
        }

    private fun JSONObject.text(key: String): String? =
        optString(key).takeIf { has(key) && !isNull(key) && it.isNotBlank() && it != "null" }
}

File: ios/Airadar/Data/BackendClient.swift
import Foundation
import UIKit

/// The Airadar backend (backend/ in this repo): flight status, accounts, trips,
/// friends, sharing, plans. Same routes the Android app uses.
enum BackendClient {

    struct BackendError: LocalizedError {
        let message: String
        let code: Int
        var errorDescription: String? { message }
    }

    // MARK: - Decoding

    static let decoder: JSONDecoder = {
        let d = JSONDecoder()
        let plain = ISO8601DateFormatter()
        plain.formatOptions = [.withInternetDateTime]
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        d.dateDecodingStrategy = .custom { decoder in
            let raw = try decoder.singleValueContainer().decode(String.self)
            if let date = fractional.date(from: raw) ?? plain.date(from: raw) { return date }
            // A naive timestamp from the server is UTC.
            if let local = LocalDateTime(string: raw) { return local.date(in: .gmt) }
            throw DecodingError.dataCorrupted(.init(codingPath: decoder.codingPath, debugDescription: "bad date \(raw)"))
        }
        return d
    }()

    static let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.dateEncodingStrategy = .iso8601
        return e
    }()

    // MARK: - Flights & airports

    static func flight(_ number: String, on day: String) async throws -> Flight {
        let path = "flights/\(number.uppercased())/\(day)"
        let data: Data
        if await AuthStore.shared.isSignedIn {
            do { data = try await authed("GET", path) } catch let e as BackendError where e.code == 401 {
                data = try await call("GET", path)
            }
        } else {
            data = try await call("GET", path)
        }
        var f = try decoder.decode(Flight.self, from: data)
        await ensureAirports(f.departure, f.arrival)
        f.callsign = f.callsign ?? AirportDatabase.shared.airlineIcao(String(number.prefix(2))).map { $0 + number.dropFirst(2) }
        return f
    }

    static func airport(_ iata: String) async throws -> Airport {
        try decoder.decode(Airport.self, from: try await call("GET", "airports/\(iata.uppercased())"))
    }

    private static func ensureAirports(_ codes: String...) async {
        for c in codes { await AirportDatabase.shared.ensure(c) { try await airport(c) } }
    }

    // MARK: - Account

    private struct TokenPair: Decodable {
        let accessToken: String
        let refreshToken: String
        let user: AuthUser
    }

    @MainActor
    static func signInWithGoogle(idToken: String) async throws -> AuthUser {
        let body = ["idToken": idToken, "device": UIDevice.current.name]
        let pair = try decoder.decode(TokenPair.self, from: try await call("POST", "auth/google", json: body))
        AuthStore.shared.save(accessToken: pair.accessToken, refreshToken: pair.refreshToken, user: pair.user)
        return pair.user
    }

    @MainActor
    static func signOut() async {
        if let token = AuthStore.shared.refreshToken {
            _ = try? await call("POST", "auth/logout", json: ["refreshToken": token])
        }
        AuthStore.shared.clear()
    }

    private static let refreshLock = AsyncLock()

    /// Swaps the refresh token for a new pair; serialised so two 401s cannot both spend it.
    private static func refreshSession() async -> Bool {
        await refreshLock.withLock {
            guard let token = await AuthStore.shared.refreshToken else { return false }
            do {
                let pair = try decoder.decode(TokenPair.self, from: try await call("POST", "auth/refresh", json: ["refreshToken": token]))
                await AuthStore.shared.save(accessToken: pair.accessToken, refreshToken: pair.refreshToken, user: pair.user)
                return true
            } catch let e as BackendError {
                if e.code == 401 { await AuthStore.shared.clear() }
                return false
            } catch { return false }
        }
    }

    private struct Me: Decodable {
        let id: String, email: String, name: String?, avatarUrl: String?, givenName: String?, color: String?
        let findableByEmail: Bool
        let membership: Membership
    }

    @MainActor
    static func me() async throws -> AuthUser {
        let m = try decoder.decode(Me.self, from: try await authed("GET", "me"))
        AuthStore.shared.updateProfile { u in
            u.givenName = m.givenName; u.color = m.color; u.findableByEmail = m.findableByEmail; u.membership = m.membership
            u.name = m.name ?? u.name; u.avatarUrl = m.avatarUrl
        }
        return AuthStore.shared.user!
    }

    @MainActor
    static func updateProfile(findableByEmail: Bool) async throws {
        let m = try decoder.decode(Me.self, from: try await authed("PATCH", "me", json: ["findableByEmail": findableByEmail]))
        AuthStore.shared.updateProfile { $0.findableByEmail = m.findableByEmail; $0.membership = m.membership }
    }

    // MARK: - Plan tokens

    private struct TokenStatus: Decodable { let plan: Tier; let until: Date; let grace: Bool; let boundEmail: String? }

    /// Before sign-in: is this token live, and may this phone use it? Binds the phone on first use.
    @MainActor
    static func checkToken(_ token: String) async throws -> CheckedToken {
        let code = token.trimmingCharacters(in: .whitespaces).lowercased()
        let s = try decoder.decode(TokenStatus.self, from: try await call("POST", "billing/token/check",
                                                                        json: ["token": code, "deviceId": AuthStore.shared.deviceId]))
        let checked = CheckedToken(token: code, tier: s.plan, until: s.until, boundEmail: s.boundEmail)
        AuthStore.shared.saveCheckedToken(checked)
        return checked
    }

    /// After sign-in: ties the token to this account (first come) and raises the plan.
    @MainActor
    static func redeem(_ token: String) async throws -> Membership {
        let m = try decoder.decode(Membership.self, from: try await authed("POST", "billing/redeem",
                                                                          json: ["token": token.lowercased(), "deviceId": AuthStore.shared.deviceId]))
        AuthStore.shared.saveMembership(m)
        return m
    }

    // MARK: - Trips

    static func listTrips(deleted: Bool = false) async throws -> [Flight] {
        let list = try decoder.decode([Flight].self, from: try await authed("GET", deleted ? "trips/deleted" : "trips"))
        for f in list { await ensureAirports(f.departure, f.arrival) }
        return list
    }

    static func putTrip(_ flight: Flight) async throws {
        var wire = flight
        wire.sharedBy = nil; wire.shares = []; wire.deletedAt = nil; wire.typicalDurationMinutes = nil
        _ = try await authed("PUT", "trips/\(flight.id)", body: try encoder.encode(wire))
    }

    static func deleteTrip(_ id: String) async throws { _ = try await authed("DELETE", "trips/\(id)") }
    static func restoreTrip(_ id: String) async throws { _ = try await authed("POST", "trips/\(id)/restore") }

    // MARK: - People and shares

    static func lookup(email: String) async throws -> Person {
        let q = email.trimmingCharacters(in: .whitespaces).addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""
        return try decoder.decode(Person.self, from: try await authed("GET", "users/lookup?email=\(q)"))
    }

    static func friends() async throws -> [Friend] { try decoder.decode([Friend].self, from: try await authed("GET", "friends")) }

    static func requestFriend(_ userId: String) async throws -> Friend {
        try decoder.decode(Friend.self, from: try await authed("POST", "friends/request", json: ["userId": userId]))
    }

    static func acceptFriend(_ id: String) async throws { _ = try await authed("POST", "friends/\(id)/accept") }
    static func removeFriend(_ id: String) async throws { _ = try await authed("DELETE", "friends/\(id)") }

    private struct ShareOut: Decodable {
        let id: String, tripId: String, status: ShareStatus, kind: String, url: String?, person: Person?
    }

    /// Shares with a friend; with no recipient a link is minted and its URL returned.
    static func shareTrip(_ tripId: String, to userId: String? = nil) async throws -> (share: TripShare?, url: String?) {
        var body: [String: Any] = [:]
        if let userId { body["toUserId"] = userId }
        let s = try decoder.decode(ShareOut.self, from: try await authed("POST", "trips/\(tripId)/share", json: body))
        let share = s.person.map { TripShare(id: s.id, person: $0, status: s.status) }
        return (share, s.url)
    }

    /// Friend shares of my trips: tripId → shares.
    static func outgoingShares() async throws -> [String: [TripShare]] {
        let rows = try decoder.decode([ShareOut].self, from: try await authed("GET", "shares/outgoing"))
        var map: [String: [TripShare]] = [:]
        for r in rows { if let p = r.person { map[r.tripId, default: []].append(TripShare(id: r.id, person: p, status: r.status)) } }
        return map
    }

    private struct IncomingShare: Decodable {
        let id: String, status: ShareStatus, person: Person?, trip: Flight
    }

    /// Friends' trips shared with me, each carrying who shared it and my answer so far.
    static func incomingShares() async throws -> [Flight] {
        let rows = try decoder.decode([IncomingShare].self, from: try await authed("GET", "shares/incoming"))
        var out: [Flight] = []
        for r in rows {
            guard let p = r.person else { continue }
            await ensureAirports(r.trip.departure, r.trip.arrival)
            var f = r.trip
            f.id = "shared:\(r.id)"
            f.sharedBy = TripShare(id: r.id, person: p, status: r.status)
            f.deletedAt = nil
            out.append(f)
        }
        return out
    }

    static func respondToShare(_ id: String, action: String) async throws {
        _ = try await authed("POST", "shares/\(id)/respond", json: ["action": action])
    }

    struct LinkedTrip: Sendable { let flight: Flight; let owner: Person; let ownerName: String }
    private struct LinkedOut: Decodable { let trip: Flight; let owner: Person; let ownerName: String? }

    /// What a shared link points at; needs no sign-in.
    static func linkedTrip(_ token: String) async throws -> LinkedTrip {
        let o = try decoder.decode(LinkedOut.self, from: try await call("GET", "shares/link/\(token)"))
        await ensureAirports(o.trip.departure, o.trip.arrival)
        return LinkedTrip(flight: o.trip, owner: o.owner, ownerName: o.ownerName ?? o.owner.givenName)
    }

    static func copyLinkedTrip(_ token: String) async throws -> Flight {
        try decoder.decode(Flight.self, from: try await authed("POST", "shares/link/\(token)/copy"))
    }

    // MARK: - Transport

    /// A signed-in call: retried once with a fresh access token after a 401.
    private static func authed(_ method: String, _ path: String, json: [String: Any]? = nil, body: Data? = nil) async throws -> Data {
        guard let first = await AuthStore.shared.accessToken else { throw BackendError(message: "Sign in required.", code: 401) }
        do {
            return try await call(method, path, json: json, body: body, bearer: first)
        } catch let e as BackendError where e.code == 401 {
            guard await refreshSession(), let next = await AuthStore.shared.accessToken else { throw e }
            return try await call(method, path, json: json, body: body, bearer: next)
        }
    }

    private static func call(_ method: String, _ path: String, json: [String: Any]? = nil, body: Data? = nil, bearer: String? = nil) async throws -> Data {
        guard Config.isBackendConfigured, let url = URL(string: "\(Config.backendURL)/\(path)") else {
            throw BackendError(message: "Backend address missing — set BACKEND_URL in Config.xcconfig.", code: 0)
        }
        var req = URLRequest(url: url, timeoutInterval: 30)
        req.httpMethod = method
        if !Config.backendToken.isEmpty { req.setValue(Config.backendToken, forHTTPHeaderField: "X-Airadar-Token") }
        if let bearer { req.setValue("Bearer \(bearer)", forHTTPHeaderField: "Authorization") }
        if let json {
            req.httpBody = try JSONSerialization.data(withJSONObject: json)
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        } else if let body {
            req.httpBody = body
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }

        let (data, response): (Data, URLResponse)
        do { (data, response) = try await URLSession.shared.data(for: req) } catch {
            throw BackendError(message: "Could not reach the Airadar server: \(error.localizedDescription)", code: 0)
        }
        let code = (response as? HTTPURLResponse)?.statusCode ?? 0
        if (200..<300).contains(code) { return data.isEmpty ? Data("{}".utf8) : data }

        // FastAPI wraps errors as {"detail": ...}; ours carry an object with "error".
        var reason = ""
        if let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            if let d = obj["detail"] as? [String: Any] { reason = (d["error"] as? String) ?? "\(d)" }
            else if let d = obj["detail"] as? String { reason = d }
        }
        let message: String
        switch code {
        case 401: message = reason.isEmpty ? "The server rejected this build's token (BACKEND_TOKEN)." : reason
        case 402: message = reason.isEmpty ? "This needs a higher plan." : reason
        case 403: message = reason.isEmpty ? "Not allowed." : reason
        case 404: message = path.hasPrefix("flights/") ? "Nothing found for that flight on that date." : (reason.isEmpty ? "Not found." : reason)
        case 429: message = "AirLabs monthly quota used up on the server."
        default: message = reason.isEmpty ? "Airadar server error (HTTP \(code))." : reason
        }
        throw BackendError(message: message, code: code)
    }
}

/// An actor-backed mutex for the one place two tasks must not overlap.
actor AsyncLock {
    private var busy = false
    private var waiters: [CheckedContinuation<Void, Never>] = []

    func withLock<T>(_ body: () async -> T) async -> T {
        if busy { await withCheckedContinuation { waiters.append($0) } }
        busy = true
        defer {
            if let next = waiters.first { waiters.removeFirst(); next.resume() } else { busy = false }
        }
        return await body()
    }
}
